File: Data Structures & Algorithms/count-paths/submission-21.java
class Solution {
    int ROWS,COLS;
    public int uniquePaths(int m, int n) {
        int[] grid = new int[n+1];
        Arrays.fill(grid,0);
        for(int i = m-1;i>=0;i--){
            int[] row = new int[n+1];
            if(i==m-1)
                row[n-1] = 1;
            for(int j = n-1;j>=0;j--){
                row[j] += grid[j]+row[j+1];
            }
            grid = row;
        }
        return grid[0];
    }

    // public int dfs(int i,int j){
    //     if(i==ROWS-1 && j == COLS-1){
    //         return 1;
    //     }
    //     if(i<0 || i==ROWS || j<0 || j==COLS){
    //         return 0;
    //     }
    //     if(grid[i][j] != null){
    //         return grid[i][j];
    //     }
    //     grid[i][j] = dfs(i+1,j) + dfs(i,j+1);
    //     return grid[i][j];
    // }
}

File: Data Structures & Algorithms/count-subsequences/submission-21.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        M,N = len(s),len(t)
        dp = [0]*(N+1)
        for i in range(M,-1,-1):
            cur = [0]*(N+1)
            for j in range(N,-1,-1):
                if j ==N:
                    cur[j] = 1
                elif i==M:
                    cur[j] = 0
                elif s[i] == t[j]:
                    cur[j] = dp[j] + dp[j+1]
                else:
                    cur[j] = dp[j]
            dp = cur
        return dp[0]


        # def dfs(i,j):
        #     if j ==N:
        #         return 1
        #     if i==M:
        #         return 0
        #     if (i,j) in dp:
        #         return dp[(i,j)]
        #     dp[(i,j)] = dfs(i+1,j)
        #     if s[i] == t[j]:
        #         dp[(i,j)] += dfs(i+1,j+1)
        #     return dp[(i,j)]  
        # return dfs(0,0)          
    
        
        